File: src/main/java/io/vertx/ext/service/discovery/DiscoveryService.java
package io.vertx.ext.service.discovery;

import io.vertx.codegen.annotations.GenIgnore;
import io.vertx.codegen.annotations.ProxyGen;
import io.vertx.codegen.annotations.VertxGen;
import io.vertx.core.AsyncResult;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.service.discovery.utils.ClassLoaderUtils;
import io.vertx.serviceproxy.ProxyHelper;

import java.util.List;

/**
 * @author <a href="http://escoffier.me">Clement Escoffier</a>
 */
@ProxyGen
@VertxGen
public interface DiscoveryService {

  void publish(String serviceItf, JsonObject properties, String address, Handler<AsyncResult<String>> resultHandler);

  void unpublish(String id, Handler<AsyncResult<Void>> resultHandler);

  void getService(String serviceItf, JsonObject filter, Handler<AsyncResult<JsonObject>> resultHandler);

  void getServices(String serviceItf, JsonObject filter, Handler<AsyncResult<List<JsonObject>>> resultHandler);

  @GenIgnore
  static <T> T getService(Class<T> itf, Vertx vertx, JsonObject reference) {
    return ProxyHelper.createProxy(itf, vertx, reference.getString("service.address"));
  }

  @GenIgnore
  static <T> T getService(Vertx vertx, JsonObject reference) {
    return getService(vertx, reference, Thread.currentThread().getContextClassLoader());
  }

  @GenIgnore
  static <T> T getService(Vertx vertx, JsonObject reference, ClassLoader classLoader) {
    Class<T> itf = ClassLoaderUtils.load(reference.getString("service.interface"), classLoader);
    if (itf == null) {
      throw new IllegalStateException("Cannot load class " + reference.getString("service.interface"));
    } else {
      return getService(itf, vertx, reference);
    }
  }

}

File: src/main/resources/vertx-service-discovery-js/discovery_service.js
/*
 * Copyright 2014 Red Hat, Inc.
 *
 * Red Hat licenses this file to you under the Apache License, version 2.0
 * (the "License"); you may not use this file except in compliance with the
 * License.  You may obtain a copy of the License at:
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */

/** @module vertx-service-discovery-js/discovery_service */
var utils = require('vertx-js/util/utils');

var io = Packages.io;
var JsonObject = io.vertx.core.json.JsonObject;
var JDiscoveryService = io.vertx.ext.service.discovery.DiscoveryService;

/**

 @class
*/
var DiscoveryService = function(j_val) {

  var j_discoveryService = j_val;
  var that = this;

  /**

   @public
   @param serviceItf {string} 
   @param properties {Object} 
   @param address {string} 
   @param resultHandler {function} 
   */
  this.publish = function(serviceItf, properties, address, resultHandler) {
    var __args = arguments;
    if (__args.length === 4 && typeof __args[0] === 'string' && (typeof __args[1] === 'object' && __args[1] != null) && typeof __args[2] === 'string' && typeof __args[3] === 'function') {
      j_discoveryService["publish(java.lang.String,io.vertx.core.json.JsonObject,java.lang.String,io.vertx.core.Handler)"](serviceItf, utils.convParamJsonObject(properties), address, function(ar) {
      if (ar.succeeded()) {
        resultHandler(ar.result(), null);
      } else {
        resultHandler(null, ar.cause());
      }
    });
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param id {string} 
   @param resultHandler {function} 
   */
  this.unpublish = function(id, resultHandler) {
    var __args = arguments;
    if (__args.length === 2 && typeof __args[0] === 'string' && typeof __args[1] === 'function') {
      j_discoveryService["unpublish(java.lang.String,io.vertx.core.Handler)"](id, function(ar) {
      if (ar.succeeded()) {
        resultHandler(null, null);
      } else {
        resultHandler(null, ar.cause());
      }
    });
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param serviceItf {string} 
   @param filter {Object} 
   @param resultHandler {function} 
   */
  this.getService = function(serviceItf, filter, resultHandler) {
    var __args = arguments;
    if (__args.length === 3 && typeof __args[0] === 'string' && (typeof __args[1] === 'object' && __args[1] != null) && typeof __args[2] === 'function') {
      j_discoveryService["getService(java.lang.String,io.vertx.core.json.JsonObject,io.vertx.core.Handler)"](serviceItf, utils.convParamJsonObject(filter), function(ar) {
      if (ar.succeeded()) {
        resultHandler(utils.convReturnJson(ar.result()), null);
      } else {
        resultHandler(null, ar.cause());
      }
    });
    } else throw new TypeError('function invoked with invalid arguments');
  };

  /**

   @public
   @param serviceItf {string} 
   @param filter {Object} 
   @param resultHandler {function} 
   */
  this.getServices = function(serviceItf, filter, resultHandler) {
    var __args = arguments;
    if (__args.length === 3 && typeof __args[0] === 'string' && (typeof __args[1] === 'object' && __args[1] != null) && typeof __args[2] === 'function') {
      j_discoveryService["getServices(java.lang.String,io.vertx.core.json.JsonObject,io.vertx.core.Handler)"](serviceItf, utils.convParamJsonObject(filter), function(ar) {
      if (ar.succeeded()) {
        resultHandler(utils.convReturnListSetJson(ar.result()), null);
      } else {
        resultHandler(null, ar.cause());
      }
    });
    } else throw new TypeError('function invoked with invalid arguments');
  };

  // A reference to the underlying Java delegate
  // NOTE! This is an internal API and must not be used in user code.
  // If you rely on this property your code is likely to break if we change it / remove it without warning.
  this._jdel = j_discoveryService;
};

// We export the Constructor function
module.exports = DiscoveryService;